### PR TITLE
[SYCLomatic] Implement help function for reduction, binary, convolution, batchnormalization

### PR DESCRIPTION
**Pull request by intwanghao**
Signed-off-by: Hao Wang <hao3.wang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 1]
```diff
@@ -52,77 +55,96 @@ class memory_desc_ext {
...
+  /// given memory descriptor.
+  /// \param [out] dt Data type.
+  /// \param [out] ndims Dimension of the memory.
+  /// \param [out] dims  Array of dimension requested_ndims that contain the 
+  /// size of each memory dimension.
+  /// \param [out] strides Array of dimension requested_ndims that contain the
+  /// stride of each memory dimension.
+  void get(int requested_ndims, dpct::library_data_t *dt, int *ndims,
+           int dims[], int strides[]) const;
+  static size_t get_element_num(memory_desc_ext desc) {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: how about move it to member function?

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 2]
```diff
@@ -853,11 +1324,48 @@ engine_ext::compress_spatial_dimensions_to_channel(::dnnl::memory::desc desc) {
     compressed_dims[1] = compressed_dims[1] * desc.data.dims[index];
     compressed_dims[index] = 1;
   }
-  return compressed_dims;
+  if (ndims == 5) {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: how to process ndims > 5?

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 3]
```diff
@@ -853,11 +1324,48 @@ engine_ext::compress_spatial_dimensions_to_channel(::dnnl::memory::desc desc) {
...
+    for (int index = 1; index < ndims; index++) {
+      channel_number = channel_number * desc.data.dims[index];
+    }
+  }
+  return ::dnnl::memory::desc({channel_number}, desc.data_type(),
+                              ::dnnl::memory::format_tag::a);
+}
+
+bool engine_ext::scale_parameter_preprocess(
+    std::vector<std::tuple<float, float, memory_desc_ext, void *>> args) {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: how about use a struct for the tuple?

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 4]
```diff
@@ -399,56 +577,132 @@ class engine_ext {
...
+
+  bool scale_parameter_preprocess(
+      std::vector<std::tuple<float, float, memory_desc_ext, void *>> args);
+
+  template <typename primitive_type, typename args_type>
+  sycl::event execute_primitive(
+      primitive_type &&primitive, args_type &&args,
+      std::vector<std::tuple<float, float, int, memory_desc_ext, void *>>
+          extra_args);
+  template <typename T> struct no_zero_op {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc:701]: how about move the struct into function transform_no_zero?

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 5]
```diff
@@ -399,56 +577,132 @@ class engine_ext {
...
+                     (int8_t *)src, (int8_t *)src + element_num, (int8_t *)dst,
+                     no_zero_op<int8_t>());
+      break;
+    case ::dnnl::memory::data_type::u8:
+      std::transform(oneapi::dpl::execution::make_device_policy(*_q),
+                     (uint8_t *)src, (uint8_t *)src + element_num,
+                     (uint8_t *)dst, no_zero_op<uint8_t>());
+      break;
+    default:
+      throw std::runtime_error("fill: unsupported data type sycl::bfloat16.");
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: bfloat16 is supported in latest compiler

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 6]
```diff
@@ -870,26 +1378,55 @@ engine_ext::compress_spatial_dimensions_to_channel(::dnnl::memory::desc desc) {
...
+    primitive_type &&primitive, args_type &&args,
+    std::vector<std::tuple<float, float, int, memory_desc_ext, void *>>
+        extra_args) {
+  std::map<void *, void *> caches_map;
+  sycl::event final_event, final_cache_event;
+
+  for (auto &extra_arg : extra_args) {
+    float alpha = std::get<0>(extra_arg);
+    float beta = std::get<1>(extra_arg);
+    int arg_index = std::get<2>(extra_arg);
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: The name arg_index is confused

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 7]
```diff
@@ -898,13 +1435,38 @@ primitive_type engine_ext::create_forward_primitive(args_type &&...args) {
       std::forward<args_type>(args)...));
 }
 
+template <bool use_attr, typename primitive_type, typename attr_type,
+          typename... args_type>
+primitive_type engine_ext::create_forward_primitive(attr_type &&attr,
+                                                    args_type &&...args) {
+  if (use_attr) {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: if constexpr (use_attr)

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 8]
```diff
@@ -870,26 +1378,55 @@ engine_ext::compress_spatial_dimensions_to_channel(::dnnl::memory::desc desc) {
...
-                 ::dnnl::memory(mem_desc.get_desc(), _eng, mem)});
-    ::dnnl::sycl_interop::execute(primitive, _s, args);
-    if(alpha == 1.f) {
-      _s.wait();
+template <typename primitive_type, typename args_type>
+sycl::event engine_ext::execute_primitive(
+    primitive_type &&primitive, args_type &&args,
+    std::vector<std::tuple<float, float, int, memory_desc_ext, void *>>
+        extra_args) {
+  std::map<void *, void *> caches_map;
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: how about use std;:vector<void *> to store the allocated memory?

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 9]
```diff
@@ -898,13 +1435,38 @@ primitive_type engine_ext::create_forward_primitive(args_type &&...args) {
...
   return typename primitive_type::primitive_desc(
       {std::forward<args_type>(args)...}, _eng);
 }
 
+template <bool use_attr, typename primitive_type, typename attr_type,
+          typename... args_type>
+typename primitive_type::primitive_desc
+engine_ext::create_forward_primitive_desc(attr_type &&attr,
+                                          args_type &&...args) {
+  if (use_attr) {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: same as above

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 10]
```diff
@@ -999,57 +1556,199 @@ void engine_ext::scale(float alpha, const memory_desc_ext &src_desc, void *src)
...
   }
+  if (onednn_algorithm == ::dnnl::algorithm::eltwise_sqrt ||
+      onednn_algorithm == ::dnnl::algorithm::eltwise_linear) {
+    if (alpha_0 != 1.f) {
+      scale(alpha_0, src_desc_0, src_0);
+    }
+    alpha_1 = 1.f;
+    auto primitive = create_forward_primitive<::dnnl::eltwise_forward>(
+        ::dnnl::prop_kind::forward_inference, onednn_algorithm,
+        src_desc_0.get_desc(), -1.f, 1.f);
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: add some comments

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 11]
```diff
@@ -999,57 +1556,199 @@ void engine_ext::scale(float alpha, const memory_desc_ext &src_desc, void *src)
...
+sycl::event engine_ext::reduction(reduction_desc &desc, float alpha,
+                                  const memory_desc_ext &src_desc, void *src,
+                                  float beta, const memory_desc_ext &dst_desc,
+                                  void *dst) {
+  if (alpha == 0.f && beta == 1.f) {
+    return sycl::event();
+  }
+  float p = 2.f;
+  reduction_op op = desc.get_reduction_op();
+  ::dnnl::algorithm onednn_algorithm;
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: how about use code like:
cache = nullptr;
switch () {
  case ...:
     cache = allocate();
     do something;
     src = cache;
     break;
   }
   event = execute();
   if (cache)
     async_dpct_free(...)

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 12]
```diff
@@ -853,11 +1324,48 @@ engine_ext::compress_spatial_dimensions_to_channel(::dnnl::memory::desc desc) {
...
+  } else {
+    for (int index = 1; index < ndims; index++) {
+      channel_number = channel_number * desc.data.dims[index];
+    }
+  }
+  return ::dnnl::memory::desc({channel_number}, desc.data_type(),
+                              ::dnnl::memory::format_tag::a);
+}
+
+bool engine_ext::scale_parameter_preprocess(
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc:1899]: add some comments

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 13]
```diff
@@ -19,6 +19,9 @@
 #include <CL/sycl.hpp>
 #include <oneapi/dnnl/dnnl.hpp>
 #include <oneapi/dnnl/dnnl_sycl.hpp>
+#include <oneapi/dpl/algorithm>
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc:19]: Should insert before <CL/sycl.hpp>, double check with Andy.

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 14]
```diff
@@ -1381,54 +1940,346 @@ bool engine_ext::scale_parameter_preprocess(
...
+    }
+
+    batch_normalization_forward_internal(
+        true, mode, epsilon, 0.f, 1.f, src_desc, src, 0.f, diff_dst_desc,
+        dst_cache, diff_scale_bias_desc, scale, bias ? bias : reordered_bias,
+        mean_var_desc, reordered_saved_mean, reordered_saved_var, nullptr,
+        nullptr);
+
+    caches.push_back(dst_cache);
+    caches.push_back(reordered_saved_mean);
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: move it into if stmt at line:2091

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 15]
```diff
@@ -1694,25 +2554,9 @@ sycl::event engine_ext::reduction(reduction_desc &desc, float alpha,
...
       onednn_algorithm, src_desc.get_desc(), dst_desc.get_desc(), p, 0.f);
 
-  if (op == reduction_op::amax) {
-    auto cache = allocate(src_desc);
-    activation_desc adesc;
-    adesc.set_algorithm(::dnnl::algorithm::eltwise_abs);
-    activation_forward(adesc, 1.f, src_desc, src, 0.f, src_desc, cache);
+  if (op == reduction_op::amax || op == reduction_op::mul_no_zeros) {
     std::unordered_map<int, ::dnnl::memory> execution_args = {
         {DNNL_ARG_SRC, ::dnnl::memory(src_desc.get_desc(), _eng, cache)}};
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: execution stmt can be moved outside if.

**Comment by LongZhenhai** [comment]
comment test

**Review by YuriPlyakhin — DISMISSED** [review]
Please, split formatting changes and new help function implementation into 2 separate pull requests.

**Comment by intwanghao** [comment]
> Please, split formatting changes and new help function implementation into 2 separate pull requests.

Done

**Pull request closed without merging**